pip/pip-304.md

```diff
@@ -0,0 +1,85 @@
+# Motivation
+
+We have a scenario that uses the `org.apache.pulsar.broker.intercept.BrokerInterceptor#messageProduced`, and need to
+obtain the `org.apache.pulsar.common.api.proto.MessageMetadata`, but the `org.apache.pulsar.broker.service.Topic.PublishContext` doesn't provide this data.
+
+I will obtain the following data from the `org.apache.pulsar.common.api.proto.MessageMetadata`:
+
+- getPublishTime
+- getEventTime
+- getDeliverAtTime
+- getPartitionKey
+
+# Goals
+
+## In Scope
+
+Make `org.apache.pulsar.broker.service.Topic.PublishContext` to provide the `org.apache.pulsar.common.api.proto.MessageMetadata`.
```

Comment:
The message metadata is not a part of `PublishContext`, we'd better not add the message metadata or headersAndPayload to the publish context because they have different recycling time. The headersAndPayload can be recycled after it has been written to bookies, don't need to wait for producer's receipt. But the PublishContext is for the producer's receipt. I think it also should be the reason why we don't add headersAndPayload to the method `BrokerInterceptor#messageProduced` but add it to `BrokerInterceptor#onMessagePublish`.

Reply:
> The headersAndPayload can be recycled after it has been written to bookies, don't need to wait for producer's receipt. 
I tried to parse the headersAndPayload in the `org.apache.pulsar.broker.service.Producer.MessagePublishContext#run`, and it works fine.

If the headersAndPayload was recycled after it has been written to bookies, it means we can parse the headersAndPayload, is it right?

Reply:
But this does not have a compatibility guarantee. A performance optimization in the future may cause the buffer to be released early, and relying on the buffer here will make it almost difficult to optimize in the future. We need to go back to the essence of the problem(Should publish context rely on the buffer) and think about it. If we shouldn't rely on the buffer but now is false, we should improve it, instead of continuing in the opposite direction